pkg/asset/machines/aws/worker.go

```diff
@@ -2,98 +2,124 @@
 package aws
 
 import (
-	"text/template"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	awsprovider "sigs.k8s.io/cluster-api-provider-aws/pkg/apis/awsproviderconfig/v1alpha1"
+	clusterapi "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
 
 	"github.com/openshift/installer/pkg/types"
 )
 
-// WorkerConfig is used to generate the worker machinesets.
-type WorkerConfig struct {
-	Instances []WorkerMachinesetInstance
-	MachineConfig
-}
+// NonMasterMachineSets returns a list of machinesets for a machinepool.
+func NonMasterMachineSets(config *types.InstallConfig, pool *types.MachinePool, userdatasecret string) ([]clusterapi.MachineSet, error) {
+	var out []clusterapi.MachineSet
+	if pool.Name == "master" {
+		return nil, fmt.Errorf("pool name master is not supported")
+	}
+	if config.Platform.AWS == nil || pool.Platform.AWS == nil {
+		return nil, fmt.Errorf("invalid inputs")
+	}
+	clustername := config.ObjectMeta.Name
+	platform := config.Platform.AWS
+	mpool := pool.Platform.AWS
+	azs := mpool.Zones
 
-// WorkerMachinesetInstance constains information specific to
-// one worker machineset.
-type WorkerMachinesetInstance struct {
-	Replicas         int64
-	AvailabilityZone string
-}
+	total := int64(0)
+	if pool.Replicas != nil {
+		total = *pool.Replicas
+	}
+	numOfAZs := int64(len(azs))
+	for idx, az := range azs {
+		replicas := int32(total / numOfAZs)
+		if int64(idx) < total%numOfAZs {
+			replicas++
+		}
 
-// MachineConfig contains fields common to worker and master
-// machine configurations
-type MachineConfig struct {
-	ClusterName string
-	AMIID       string
-	Tags        map[string]string
-	Region      string
-	Machine     types.AWSMachinePoolPlatform
-}
+		tags := map[string]string{
+			"tectonicClusterID":                                  config.ClusterID,
+			fmt.Sprintf("kubernetes.io/cluster/%s", clustername): "owned",
+		}
+		for k, v := range platform.UserTags {
+			tags[k] = v
+		}
+		ptags := []awsprovider.TagSpecification{}
+		for k, v := range tags {
+			ptags = append(ptags, awsprovider.TagSpecification{Name: k, Value: v})
+		}
+
+		provider := awsprovider.AWSMachineProviderConfig{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "aws.cluster.k8s.io/v1alpha1",
+				Kind:       "AWSMachineProviderConfig",
+			},
+			InstanceType:       mpool.InstanceType,
+			AMI:                awsprovider.AWSResourceReference{ID: &mpool.AMIID},
+			Tags:               ptags,
+			IAMInstanceProfile: &awsprovider.AWSResourceReference{ID: func(s string) *string { return &s }(fmt.Sprintf("%s-worker-profile", clustername))},
```

Comment:
Curious should the machine pool's IAMRoleName be used somewhere in this code?

Reply:
Didn't find any place to put the IAM role in the machine definition. do you have an idea?